File: ruby/inorder_successor.rb
def inorder_successor(root, p)
    return nil if root.nil?
    queue = [ root ]
    min = Float::INFINITY
    max = nil
    while queue.length > 0
        node = queue.shift
        
        if node.val > p
            min = [min, node.val].min
        end
        
        if node.val == p
            max = node.val
            break
        end
        
        queue.push(node.left) if !node.left.nil?
        queue.push(node.right) if !node.right.nil?
    end
    
    return min if min != Float::INFINITY && max != nil
    return nil
end

class Node
    attr_accessor :val, :left, :right
    def initialize(val)
        self.val = val;
        self.left = self.right = nil
    end
end

 node1 = Node.new(2)
 node2 = Node.new(1)
 node3 = Node.new(3)
#  node4 = Node.new(2)
#  node5 = Node.new(4)
#  node6 = Node.new(1)
 
 node1.left = node2
 node1.right = node3
#  node2.left = node4
#  node2.right = node5
#  node4.left = node6

p inorder_successor(node1, 1)

File: inorder_successor.js
var inorderSuccessor = function(root, p) {
    let successor = Infinity;
    let queue = [root];
    while (queue.length) {
        let node = queue.shift();
        
        if (node.val > p) successor = Math.min(successor, node.val);
        if (node.left) queue.push(node.left);
        if (node.right) queue.push(node.right);
    }
    return successor === Infinity ? null : successor
};

class Node{
    constructor(val) {
        this.left = this.right = null;
        this.val = val;
    }
}

let node1 = new Node(2);
let node2 = new Node(1);
let node3 = new Node(3);

console.log(inorderSuccessor(node1, 1))